std/stdio.d

```diff
@@ -4043,35 +4082,40 @@ private struct ReadlnAppender
     import core.stdc.string;
 
     char[] buf;
-    size_t cap;
     size_t pos;
-    size_t initcap;
+    bool newBuf = false;
 
     void initialize(char[] b)
     {
         buf = b;
-        cap = initcap = buf.capacity;
-        if (cap < buf.length)
-            cap = buf.length;
         pos = 0;
     }
     @property char[] data()
     {
-        // always shrink/extend the buffer to allow reuse in the next call
-        if (initcap > 0 || cap > buf.length)
+        /* When buf is not known to the outside world, it's safe to append to
+        the returned slice. */
+        if (newBuf)
             assumeSafeAppend(buf.ptr[0..pos]);
         return buf.ptr[0..pos];
     }
 
     void reserve(size_t n)
     {
-        if (pos + n > cap)
+        if (buf.length >= pos + n) // buf is already large enough
+            return;
+
+        immutable curCap = buf.capacity;
+        if (curCap >= pos + n)
+        {
+            buf.length = curCap;
```

Comment:
Ok, I think I got it. In this scenario, the slice that's returned from `data` is guaranteed to be longer than the originally passed buffer. So it's safe to say that any capacity we didn't use is free for all. Amending.